src/main.cpp

```diff
@@ -1045,19 +1052,125 @@ bool ContextualCheckTransaction(
+            if (tx.GetConsensusBranchId()) {
+                return state.DoS(
+                    dosLevelPotentiallyRelaxing,
+                    error("ContextualCheckTransaction(): pre-NU5 transaction has consensus branch id set"),
+                    REJECT_INVALID, "bad-tx-has-consensus-branch-id");
+            }
+        }
+
+        // Reject transactions with invalid version
+        if (tx.nVersionGroupId == ZIP225_VERSION_GROUP_ID) {
+            if (tx.nVersion < ZIP225_MIN_TX_VERSION) {
+                return state.DoS(
+                    dosLevelConstricting,
+                    error("ContextualCheckTransaction(): ZIP225 version too low"),
+                    REJECT_INVALID, "bad-tx-zip225-version-too-low");
+            }
+
+            if (tx.nVersion > ZIP225_MAX_TX_VERSION) {
+                return state.DoS(
+                    dosLevelPotentiallyRelaxing,
+                    error("ContextualCheckTransaction(): ZIP225 version too high"),
+                    REJECT_INVALID, "bad-tx-zip225-version-too-high");
+            }
+
+            // tx.nConsensusBranchId must match the current consensus branch id
+            if (!(tx.GetConsensusBranchId() && *tx.GetConsensusBranchId() == consensusBranchId)) {
+                return state.DoS(
+                    dosLevelPotentiallyRelaxing,
+                    error("ContextualCheckTransaction(): consensus branch id not set"),
+                    REJECT_INVALID, "bad-tx-missing-consensus-branch-id");
+            }
+        }
+
+        //nSpendsSapling, nOutputsSapling, and nActionsOrchard MUST all be less than 2^16
+        size_t max_spends = 1 << 16;
+        if (tx.vShieldedSpend.size() >= max_spends) {
+            return state.DoS(
+                dosLevelPotentiallyRelaxing,
+                error("ContextualCheckTransaction(): More than 2^16 Sapling spends"),
+                REJECT_INVALID, "bad-tx-too-many-sapling-spends");
+        }
+        if (tx.vShieldedOutput.size() >= max_spends) {
+            return state.DoS(
+                dosLevelPotentiallyRelaxing,
+                error("ContextualCheckTransaction(): More than 2^16 Sapling outputs"),
+                REJECT_INVALID, "bad-tx-too-many-sapling-outputs");
+        }
+        if (orchard_bundle.GetNumActions() >= max_spends) {
+            return state.DoS(
+                dosLevelPotentiallyRelaxing,
+                error("ContextualCheckTransaction(): More than 2^16 Orchard actions"),
+                REJECT_INVALID, "bad-tx-too-many-orchard-actions");
+        }
+
+        if (tx.IsCoinBase()) {
+            // A coinbase transaction cannot have Orchard spends.
+            if (orchard_bundle.SpendsEnabled()) {
+                return state.DoS(
+                    dosLevelPotentiallyRelaxing,
+                    error("ContextualCheckTransaction(): coinbase has orchard shielded spends"),
+                    REJECT_INVALID, "bad-cb-has-orchard-spends");
+            }
+
```

Comment:
This is being checked non-contextually (in `CheckTransactionWithoutProofVerification` below).
```suggestion
```